dumpState Improvements

Ashod opened the pull request:
Better formatting of `dumpState()` and more memory stats included. Also, we now forward SIGUSR1 to the child processes, to get their memory stats and other info.

- wsd: better dumpState formatting
- wsd: dump malloc-info with USR1
- wsd: refactor forwarding a signal to our children
- wsd: forward SIGUSR1 to children
- wsd: unify dump_malloc_state with getMallocInfo
- wsd: dump malloc_info in the children

mmeeks commented on net/HttpRequest.hpp:
```diff
@@ -755,9 +755,9 @@ class Request final
-        os << indent << "stage: " << name(_stage);
-        os << indent << "headers: ";
-        _header.serialize(os);
+        os << indent << "\tstage: " << name(_stage);
```
Curious as to why we're adding another indent to indent in-line here, rather than just bumping that in the parent caller ?

Ashod replied:
It makes it more readable to have the indents:
```
http::server::Session: #123 
    connected: ...
    startTime: ...
```

Rather than have them all on the same level. It's just a matter of indenting the members of the object vs it's title, for readability (these dumps do get very long and tiring/confusing to read).

mmeeks replied:
Not a fan of the large number of lines changed; but hey.

mmeeks commented on common/Util.cpp:
```diff
@@ -806,6 +807,29 @@ namespace Util
         std::_Exit(code);
     }
 
+    std::string getMallocInfo()
```
Like it =)

mmeeks commented on wsd/COOLWSD.cpp:
```diff
@@ -4976,20 +4978,30 @@ void forwardSigUsr2()
+    forwardSignal(SIGUSR2);
+}
+
+void forwardSignal(const int signum)
```
Nice minimal patch; good stuff.

mmeeks commented on wsd/COOLWSD.cpp:
```diff
@@ -4959,6 +4959,10 @@ void dump_state()
     Admin::dumpMetrics();
 #endif
+
+    std::lock_guard<std::mutex> docBrokerLock(DocBrokersMutex);
```
I assume and hope that this lock ordering is consistent everywhere =)

Ashod replied:
I maintained the same order, yes. And it seems these signal-triggered functions are the only place we need to take both locks. I didn't revise the logic, though not a fan of multi-locks.

mmeeks reviewed (changes requested):
Few cleanups / squashes here needed - but looks good. Thanks Ash =)

Ashod commented on common/SigUtil.cpp:
```diff
@@ -159,6 +159,68 @@ void requestShutdown()
         UnattendedRun = true;
     }
 
+    const char *signalName(const int signo)
```
I didn't want to add `#if !MOBILEAPP`. That's the only reason. I really hated moving the large code (and the 8-year-old git blame). So happy to replace it with ifdefs.

Ashod commented:
Thanks, @mmeeks, for the feedback. Updated per request and added a new patch to include buffer/cache capacities (in addition to sizes), to gauge the memory footprint of these buffers/caches.

mmeeks reviewed (approved):
Looks ok - but what are all these horror MOBILEAPP conditionals doing? surely we can have no-op methods inside if we want that return dummy values and will be GC'd out by the linker on mobile - without the conditionals (?)

Otherwise lets get this in then for green then fix it further.

Ashod commented on common/SigUtil.cpp:
```diff
@@ -232,6 +232,7 @@ void requestShutdown()
         signalLog(buf + i + 1);
     }
 
+#endif // !MOBILEAPP
```
But that's what moving this function tried to avoid. The function is blocked in `#if !MOBILEAPP`, and we need to call on mobile, so we either move it out, or we exclude it, like I do here.
